src/algebra/discrete-log.md

```diff
@@ -113,32 +109,96 @@ This can be done by keeping a variable that is multiplied by $a$ each time we in
 With this change, the complexity of the algorithm is still the same, but now the $\log$ factor is only for the `map`.
 Instead of a `map`, we can also use a hash table (`unordered_map` in C++) which has the average time complexity $O(1)$ for inserting and searching.
 
-```cpp
+Problems often ask for the minimum $x$ which satisfies the solution.  
+It is possible to get all answers and take the minimum, or reduce the first found answer using [Euler's theorem](./algebra/phi-function.html#toc-tgt-2), but we can be smart about the order in which we calculate values and ensure the first answer we find is the minimum.
+
+```cpp discrete_log
+// Returns minimum x for which a ^ x % m = b % m, a and m are coprime.
 int solve(int a, int b, int m) {
+	a %= m, b %= m;
 	int n = (int) sqrt (m + .0) + 1;
 
 	int an = 1;
 	for (int i = 0; i < n; ++i)
 		an = (an * 1ll * a) % m;
 
 	map<int, int> vals;
-	for (int p = 1, cur = an; p <= n; ++p) {
-		if (!vals.count(cur))
-			vals[cur] = p;
+	for (int q = 0, cur = b; q <= n; ++q) {
+		vals[cur] = q;
+		cur = (cur * 1ll * a) % m;
+	}
+
+	for (int p = 1, cur = 1; p <= n; ++p) {
 		cur = (cur * 1ll * an) % m;
+		if (vals.count(cur)) {
+			int ans = n * p - vals[cur];
+			return ans;
+		}
 	}
+	return -1;
+}
```

Comment:
The implementation does not cover all the cases. Check the following cases:

```
a=2, b=3, m=4, x should be -1
a=2, b=6, m=10, x should be 4
a=3589, b=58, m=97, x should be -1
```

In all of the above cases, the proposed method returns `0`